insights/parsers/ls.py

```diff
@@ -441,3 +444,157 @@ def parse_content(self, content):
+        True
+        >>> ls_files_se.listing_dirs
+        ['/boot']
+        >>> "/dev/vda" in ls_files_se.listing_specials
+        True
+        >>> ls_files_se.entries['/boot']['perms']
+        'r-xr-xr-x.'
+    """
+
+    def parse_content(self, content):
+        """
+        Parses a file/directory listing with SELinux information from the
+        provided content and updates the dictionary with all parsed entries.
+        """
+        self.update(ls_parser.parse(content, None)[None])
+
+    @property
+    def listing_files(self):
+        """
+        The list of all non-special files (i.e. not block or character files).
+        """
+        return self['files']
+
+    @property
+    def listing_dirs(self):
+        """
+        The list of all directories.
+        """
+        return self['dirs']
+
+    @property
+    def entries(self):
+        """
+        The listing of all entries.
+        Entries that can be parsed then have fields as described in the class
+        description above.
+
+        .. note::
+            The 'raw_entry' key is removed from the return value.  Use the
+            `raw_entry_of` method instead.
+        """
+        return self['entries']
+
+    @property
+    def listing_specials(self):
+        """
+        The list of all block and character special files.
+        """
+        return self['specials']
+
+    def raw_entry_of(self, target):
+        """
+        Returns the raw line entry of the ``target`` listed in
+        'ls' command output.
+
+        Parameters:
+            target (string): Name of the directory or file to get
+                information for.
+
+        Returns:
+            str: The re-constructed rough line if found or None if not found.
+
+        .. note::
+            As it's re-constructed according to the serialized items, it's not
+            identical with the original line.
+        """
+        entries = self['entries']
+        if target in entries:
+            tgt = entries[target]
+            raw_line = ' '.join([tgt['type'] + tgt['perms']])
+            if 'links' in tgt:
+                raw_line += ' ' + str(tgt['links'])
+            raw_line += ' ' + ' '.join([tgt['owner'], tgt['group']])
+            if 'se_user' in tgt:
+                raw_line += ' ' + ':'.join(
+                    [tgt['se_user'], tgt['se_role'], tgt['se_type'], tgt['se_mls']]
+                )
+            if 'size' in tgt:
+                raw_line += ' ' + str(tgt['size'])
+            if 'major' in tgt:
+                raw_line += ' ' + ', '.join([str(tgt['major']), str(tgt['minor'])])
+            if 'date' in tgt:
+                raw_line += ' ' + tgt['date']
+            raw_line += ' ' + tgt['name']
+            if tgt['type'] == 'l' and 'link' in tgt:
+                raw_line += ' -> ' + tgt['link']
+            return raw_line
+        return None
+
+
+@parser(Specs.ls_ldZ)
+class LSldZ(FileListingNoHeadSelinux):
```

Comment:
Just out of curious, seeing the other ls parsers are using this `FileListing`, including specs w/ or w/o the selinux content, eg. `ls_laZ` and `ls_la` .  
Is it possible to reuse this `FileListing` for `ls_ldZ`?  say, enhancing the `FileListing` functions to have a default value `None` for arg `directory`?

Reply:
As discussed by Xiangce, we will continue using FileListingNoHeadSelinux. This is because ls -ld is different from ls (it does not include subdirectories or files, and files and directories in this spec are not relevant). The name of the base class will be updated to FileListingNoHead.